control/freqplot.py

```diff
@@ -723,37 +814,101 @@ def nyquist_plot(syslist, omega=None, plot=True, omega_limits=None,
+                    p_ol = splane_poles[
+                        (np.abs(splane_poles - p_cl)).argmin()]
+
+                    if abs(p_ol - p_cl) <= indent_radius and \
+                       warn_encirclements:
+                        warnings.warn(
+                            "indented contour may miss closed loop pole; "
+                            "consider reducing indent_radius to be less than "
+                            f"{abs(p_ol - p_cl):5.2g}", stacklevel=2)
+
+            #
+            # See if we should add some frequency points near imaginary poles
+            #
+            for p in splane_poles:
+                # See if we need to process this pole (skip if on the negative
+                # imaginary axis or not near imaginary axis + user override)
+                if p.imag < 0 or abs(p.real) > indent_radius or \
+                   omega_range_given:
+                    continue
+
+                # Find the frequencies before the pole frequency
+                below_points = np.argwhere(
+                    splane_contour.imag - abs(p.imag) < -indent_radius)
+                if below_points.size > 0:
+                    first_point = below_points[-1].item()
+                    start_freq = p.imag - indent_radius
+                else:
+                    # Add the points starting at the beginning of the contour
+                    assert splane_contour[0] == 0
+                    first_point = 0
+                    start_freq = 0
+
+                # Find the frequencies after the pole frequency
+                above_points = np.argwhere(
+                    splane_contour.imag - abs(p.imag) > indent_radius)
+                last_point = above_points[0].item()
+
+                # Add points for half/quarter circle around pole frequency
+                # (these will get indented left or right below)
+                splane_contour = np.concatenate((
+                    splane_contour[0:first_point+1],
+                    (1j * np.linspace(
+                        start_freq, p.imag + indent_radius, indent_points)),
+                    splane_contour[last_point:]))
+
+            # Indent points that are too close to a pole
             for i, s in enumerate(splane_contour):
                 # Find the nearest pole
                 p = splane_poles[(np.abs(splane_poles - s)).argmin()]
+
                 # See if we need to indent around it
                 if abs(s - p) < indent_radius:
-                    if p.real < 0 or (np.isclose(p.real, 0) \
-                            and indent_direction == 'right'):
+                    # Figure out how much to offset (simple trigonometry)
+                    offset = np.sqrt(indent_radius ** 2 - (s - p).imag ** 2) \
+                        - (s - p).real
+
+                    # Figure out which way to offset the contour point
+                    if p.real < 0 or (p.real == 0 and
```

Comment:
I think I was the one who changed this code from equality to zero to is close to zero for the reason that sometimes when working with discrete systems you start with a continuous time system, with poles on imag axis, and then discretize it. When these poles are mapped back to the s plane in this code, their real part doesn’t end up to be perfectly zero anymore. 

I Probably should have written a unit test!

Reply:
The current code should OK because the Nyquist criterion is "robust" with respect to whether you indent to the right or the left.  You just have to be consistent in how you count open and closed loop poles.

Here's an example (from `tests/nyquist_test.py`):
```
sys_c = ct.tf([1], [1, 3, 2]) * ct.tf([1], [1, 0, 1])
sys = ct.c2d(sys_c, 0.01)
np.abs(sys.poles()) - 1
array([ 9.58553237e-11,  9.58553237e-11, -9.95016656e-03, -1.98013266e-02])
```
The closed loop system has poles at
```
sys.feedback().poles()
array([1.00116706+0.01054462j, 1.00116706-0.01054462j,
       0.98390719+0.00189796j, 0.98390719-0.00189796j])
```
so it is unstable (as is the continuous time system, which has a pair of poles with real part 0.122). 

If you run the current code, it will map the z-plane poles to
```
np.log(sys.poles()) / sys.dt
array([ 9.58553380e-09+1.j,  9.58553380e-09-1.j, -1.00000003e+00+0.j,
       -1.99999999e+00+0.j])
```
Note that what used to be a purely imaginary pole (at +/- 1j) is now slightly unstable (as you described).  But the Nyquist criterion is still OK: we indent to the _left_ (since they are unstable) and we end up with zero net encirclements:
```
count = ct.nyquist_plot(sys)
```
![Figure_1](https://user-images.githubusercontent.com/293362/163697990-6c0ef7a1-dee7-4e79-b97f-b3d6461db11c.png)

Since Z = N + P = 0 + 2 = 2, we conclude that the closed loop system has two poles in the right half-plane (correct!).

If it had turned out that the poles had mapped to exactly on the unit circle or to just inside the unit circle, then we would get a plot that looks like the continuous time plot for this example:
![Figure_2](https://user-images.githubusercontent.com/293362/163698076-e134ceb2-7ecc-43af-be60-8a1a0d7dbbb0.png)
Note that in this case there are 2 encirclements and we get Z = N + P = 2 + 0 = 2, which is the same as before.